### ISSUE TITLE
bug in dataframe.describe when using percentiles parameter

### ISSUE THREAD

**Issue by gregrf** [dask/dask#4642]
Think that I found a bug in `dataframe.describe`. I'm using dask 1.1.4, numpy 1.16.2, pandas 0.24.2.

For a dataframe, `df`, `df.describe(percentiles=[0.25, 0.50, 0.75])` should be the same as `df.describe()` and it is but if I add in a higher percentile like 0.99 it breaks the 75'th percentile which seems to become a copy of the 50'th percentile:

```python
A = dask.array.arange(100)
df = dask.dataframe.from_dask_array(A)
df.describe().compute()
```

count    100.000000
mean      49.500000
std       29.011492
min        0.000000
25%       24.750000
50%       49.500000
75%       74.250000
max       99.000000
dtype: float64

```python
df.describe(percentiles=[0.25, 0.50, 0.75, 0.99]).compute()
```

count    100.000000
mean      49.500000
std       29.011492
min        0.000000
25%       24.750000
50%       49.500000
99%       98.010000
75%       49.500000
max       99.000000
dtype: float64

Notice that the 75'th percentile has changed to a copy of the 50'th percentile.

**Comment by gregrf** [comment]
FWIW, the equivalent code produced the expected result in pandas.

**Comment by mrocklin** [comment]
This may be due to a failure of our approximate quantiles algorithm.  See #1225 for more information.

**Comment by gregrf** [comment]
I don't think so. With the same arguments, dataframe.quantile() doesn't appear to have this issue. 

>>> A = dask.array.arange(100)
>>> df = dask.dataframe.from_dask_array(A)
>>> df.quantile([0.25, 0.50, 0.75]).compute()

0.25    24.75
0.50    49.50
0.75    74.25
dtype: float64

>>> df.quantile([0.25, 0.50, 0.75, 0.99]).compute()

0.25    24.75
0.50    49.50
0.75    74.25
0.99    98.01
dtype: float64

The 75'th percentile has not changed like it did with describe() when I added the 99'th percentile.

**Comment by gregrf** [comment]
I think I've isolated the problem. Here's the code for dataframe.describe():

```python
@derived_from(pd.DataFrame)
def describe(self, split_every=False, percentiles=None):
    # currently, only numeric describe is supported
    num = self._get_numeric_data()
    if self.ndim == 2 and len(num.columns) == 0:
        raise ValueError("DataFrame contains only non-numeric data.")
    elif self.ndim == 1 and self.dtype == 'object':
        raise ValueError("Cannot compute ``describe`` on object dtype.")
    if percentiles is None:
        percentiles = [0.25, 0.5, 0.75]
    else:
        percentiles = list(set(sorted(percentiles + [0.5])))   # HERE!
    stats = [num.count(split_every=split_every),
             num.mean(split_every=split_every),
             num.std(split_every=split_every),
             num.min(split_every=split_every),
             num.quantile(percentiles),
             num.max(split_every=split_every)]
    stats_names = [(s._name, 0) for s in stats]

    name = 'describe--' + tokenize(self, split_every)
    layer = {(name, 0): (methods.describe_aggregate, stats_names)}
    graph = HighLevelGraph.from_collections(name, layer, dependencies=stats)
    return new_dd_object(graph, name, num._meta, divisions=[None, None])
```

when I pass `percentiles=[0.25, 0.50, 0.75, 0.99]` the percentiles get passed to the line I've marked "HERE!"

But look what happens in that line:
```python
percentiles = [0.25, 0.50, 0.75, 0.99]
percentiles = list(set(sorted(percentiles + [0.5])))
percentiles
```

[0.25, 0.5, 0.99, 0.75]

`set` doesn't produce a sorted output. `sorted` needs to be called after `set`. I'd re-write the line

```python
percentiles = [0.25, 0.50, 0.75, 0.99]
percentiles = sorted(set(percentiles + [0.5]))
percentiles
```

[0.25, 0.5, 0.75, 0.99]

**Comment by gregrf** [comment]
I issued a pull request for my change to that line: https://github.com/dask/dask/pull/4647 .

**Comment by gregrf** [comment]
That pull request just fixes the specific bug where describe sometimes shuffles the order of the percentiles which it obviously should never do. But this leaves the issue of why quantile() gives wrong results when it's quantiles argument is unsorted. For example

```python
import dask.dataframe
import dask.array
import dask.dataframe

A = dask.array.arange(100)
s = dask.dataframe.from_dask_array(A)
s.quantile([0.75, 0.50, 0.25]).compute()
```
0.75    24.75
0.50    74.25
0.25    24.75
dtype: float64

Notice that the results for the 25th and 50th percentiles are wrong here.

pandas doesn't require that the quantiles be sorted and I don't see anything in dask's documentation requiring it either. pandas calculates the quantiles correctly and keeps the order of the argument. We could reproduce that, however, it's a really simple 1-line patch to make `quantile()` handle unsorted quantiles if we just always sort the input. I recommend that as a stop-gap because the status quo of silently outputting incorrect results is the worst possible case. Here's a simple patch: https://github.com/dask/dask/pull/4650 .

**Comment by jrbourbeau** [comment]
Closing this issue as it seems to be resolved in #4647 and #4650
